lib/hash.ts

```diff
@@ -0,0 +1,120 @@
+// This module implements the "double hash" signature specification described in docs/signature_specification.md
+import * as crypto from 'crypto';
+import { SignatureResult } from './types';
+import { isBinary } from './utils/binary';
+import {
+  hasLineEndings,
+  detectLineEndingFormat,
+  transformLineEndings,
+  removeUnwantedBytes,
+} from './utils/format';
+
+const hashAlgorithm = 'md5';
+
+const regex = /=/g;
+
+enum LineEndingFormat {
+  LINUX,
+  WINDOWS,
+  UNKNOWN,
+}
+
+type DubHashResult = [string, string] | [string, null];
+
+export function getHashSignature(
+  filePath: string,
+  fileContents: Buffer,
+  getAltHash = false,
+): SignatureResult | null {
+  const fileIsBinary = isBinary(fileContents);
+  const hashes_ffm = [];
+
+  // Unless getAltHash is true, we get the default hash types
+  if (getAltHash) {
+    const [digest1, digest2] = getDubHashDigests(fileContents, fileIsBinary);
+    // Check that digest1 could be generated
+    if (!digest1) {
+      throw Error(`Failed to generate hash of ${filePath}`);
+    }
+
+    hashes_ffm.push({
+      data: digest1,
+      format: 1,
+    });
+
+    // digest2 not being generated is ok, just check that it has been generated
+    // before adding it.
+    if (digest2) {
+      hashes_ffm.push({
+        data: digest2 as string,
+        format: 1,
+      });
+    }
+  } else {
+    // Add "half double hash"
+    hashes_ffm.push({
+      data: getSingleDigest(fileContents),
+      format: 1,
+    });
+  }
+
+  // Always generate uhash
+  hashes_ffm.push({
+    data: getUhashDigest(fileContents, fileIsBinary),
+    format: 3,
+  });
+
+  return {
+    path: filePath,
+    size: fileContents.length,
+    hashes_ffm: hashes_ffm,
+  };
+}
+
+export function getDubHashDigests(
+  fileContents: Buffer,
+  isBinary: boolean,
+): DubHashResult {
+  // If the file is binary, you only produce one hash
+  // If the file does not have line endings, you treat it like a binary file
+  // If the LF character is not found, the line endings are UNKNOWN, and we produce only one hash.
+  const lineEndingFormat: LineEndingFormat = detectLineEndingFormat(
+    fileContents,
+  );
+  if (
+    isBinary ||
+    !hasLineEndings(fileContents) ||
+    lineEndingFormat === LineEndingFormat.UNKNOWN
+  ) {
+    return [getSingleDigest(fileContents), null];
+  }
+  // NB: You MUST calculate the digest for the original file and store it, since transformLineEndings
+  // will modify fileContents buffer.
+  const digestForOriginal = getSingleDigest(fileContents);
+  const transformedFile = transformLineEndings(fileContents, lineEndingFormat);
+  return [digestForOriginal, getSingleDigest(transformedFile)];
+}
+
+function getSingleDigest(fileContents: Buffer): string {
+  if (hashAlgorithm == 'md5') {
+    const hash = crypto.createHash(hashAlgorithm).update(fileContents);
+    return hash.digest('base64').replace(regex, '');
+  } else {
```

Comment:
shouldn't we throw exception for all non supported algorithms for now?
(empty string is not an error message in my opinion, even Go (who lacks exception), returns an error)